drivers/interrupt_controller/intc_clic.c

```diff
@@ -0,0 +1,322 @@
+/*
+ * Copyright (c) 2021 Tokita, Hiroshi <tokita.hiroshi@gmail.com>
+ * Copyright (c) 2025 Andes Technology Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @brief Driver for Core Local Interrupt Controller
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/arch/riscv/csr.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/interrupt_controller/riscv_clic.h>
+#include "intc_clic.h"
+
+#if DT_HAS_COMPAT_STATUS_OKAY(riscv_clic)
+#define DT_DRV_COMPAT riscv_clic
+#elif DT_HAS_COMPAT_STATUS_OKAY(nuclei_eclic)
+#define DT_DRV_COMPAT nuclei_eclic
+#endif
+
+struct clic_data {
+	uint8_t nlbits;
+	uint8_t intctlbits;
+};
+
+struct clic_config {
+	mem_addr_t base;
+};
+
+#ifdef CONFIG_PMP_STACK_GUARD
+/*
+ * M-mode CLIC memory-mapped registers are accessible only in M-mode.
+ * Temporarily disable the PMP stack guard (set mstatus.MPRV = 0) to configure
+ * CLIC registers, then restore the PMP stack guard using these macros.
+ */
+#define DISABLE_PMP_STACK_GUARD                                                                    \
+	unsigned int key = irq_lock();                                                             \
+	unsigned long _mstatus = csr_read_clear(mstatus, MSTATUS_MPRV);
+
+#define RESTORE_PMP_STACK_GUARD                                                                    \
+	csr_write(mstatus, _mstatus);                                                              \
+	irq_unlock(key);
```

Comment:
We should probably guard there macros with a `do { ... } while (0)`

Reply:
Right, thanks for your feedback.
Because the variable declaration (key, _mstatus) across two `do { ... } while (0)` macros is tricky, I changed the implementation to function call with `if (IS_ENABLED( ... ))`.